File: src/main/java/org/jcodec/containers/mp4/MPDModel.java
package org.jcodec.containers.mp4;

import java.io.IOException;
import java.net.URL;
import java.util.LinkedList;
import java.util.List;

import javax.xml.stream.XMLStreamException;

import org.jcodec.api.JCodecException;
import org.jcodec.common.XMLMapper;
import org.jcodec.common.XMLMapper.TypeHandler;
import org.jcodec.common.model.Rational;

public class MPDModel {
    public static class Time {
        public double sec;

        public Time(double sec) {
            this.sec = sec;
        }

        public static Time parseTime(String arg) {
            char[] charArray = arg.toCharArray();
            if (charArray.length < 3 || charArray[0] != 'P' || charArray[1] != 'T')
                return null;
            int tokenStart = 2;
            double sec = 0;
            for (int i = 2; i < charArray.length; i++) {
                if (charArray[i] == 'S' || charArray[i] == 'M' || charArray[i] == 'H' && tokenStart != i) {
                    String token = new String(charArray, tokenStart, i - tokenStart);
                    double parseDouble = Double.parseDouble(token);
                    tokenStart = i + 1;
                    switch (charArray[i]) {
                    case 'S':
                        sec += parseDouble;
                        break;
                    case 'M':
                        sec += parseDouble * 60;
                        break;
                    case 'H':
                        sec += parseDouble * 3600;
                        break;
                    }
                }

            }
            return new Time(sec);
        }

        public double getSec() {
            return sec;
        }
    }

    public static class TypeHandler1 implements TypeHandler {

        @Override
        public Object parse(String value, Class<?> clz) {
            if (clz == Time.class) {
                return Time.parseTime(value);
            }
            return null;
        }

        @Override
        public boolean supports(Class<?> clz) {
            return clz == Time.class;
        }
    }

    public static class MPD {
        public Time mediaPresentationDuration;
        public List<Period> periods = new LinkedList<Period>();

        public void addPeriod(Period arg) {
            periods.add(arg);
        }
    }

    public static class Period {
        public Time start;
        public Time duration;
        public List<AdaptationSet> adaptationSets = new LinkedList<AdaptationSet>();

        public void addAdaptationSet(AdaptationSet arg) {
            this.adaptationSets.add(arg);
        }
    }

    public static class AdaptationSet {
        public boolean segmentAlignment;
        public int maxWidth;
        public int maxHeight;
        public Rational maxFrameRate;
        public Rational par;

        public SegmentTemplate segmentTemplate;
        public List<Representation> representations = new LinkedList<Representation>();

        public void addRepresentation(Representation arg) {
            representations.add(arg);
        }
    }

    public static class SegmentTemplate {
        public int timescale;
        public int duration;
        public String media;
        public int startNumber;
        public String initialization;
    }

    public static class Representation {
        public String id;
        public String mimeType;
        public String codecs;
        public int width;
        public int height;
        public Rational frameRate;
        public Rational sar;
        public int startWithSAP;
        public int bandwidth;
        public int audioSamplingRate;
        public String baseURL;

        public SegmentTemplate segmentTemplate;
        public SegmentBase segmentBase;
        public SegmentList segmentList;
    }

    public static class SegmentBase {
        public String indexRange; // "0-834"
    }

    public static class SegmentList {
        public int duration;
        public Initialization initialization;
        public List<SegmentURL> segmentUrls = new LinkedList<SegmentURL>();

        public void addSegmentURL(SegmentURL arg) {
            segmentUrls.add(arg);
        }
    }

    public static class Initialization {
        public String sourceURL;
    }

    public static class SegmentURL {
        public String media;
    }

    public static MPD parse(URL url) throws IOException, JCodecException {
        try {
            return XMLMapper.map(url.openStream(), MPDModel.MPD.class, new TypeHandler1());
        } catch (ReflectiveOperationException e) {
            throw new JCodecException(e);
        } catch (XMLStreamException e) {
            throw new JCodecException(e);
        }
    }
}

File: src/main/java/org/jcodec/containers/mp4/MPDModel.kt
package org.jcodec.containers.mp4

import org.jcodec.api.JCodecException
import org.jcodec.common.XMLMapper
import org.jcodec.common.XMLMapper.TypeHandler
import org.jcodec.common.model.Rational
import java.io.IOException
import java.net.URL
import java.util.*
import javax.xml.stream.XMLStreamException

object MPDModel {
    @Throws(IOException::class, JCodecException::class)
    fun parse(url: URL): MPD {
        return try {
            XMLMapper.map(url.openStream(), MPD::class.java, TypeHandler1())
        } catch (e: ReflectiveOperationException) {
            throw JCodecException(e)
        } catch (e: XMLStreamException) {
            throw JCodecException(e)
        } catch (e: Exception) {
            throw JCodecException(e)
        }
    }

    class Time(var sec: Double) {

        companion object {
            fun parseTime(arg: String): Time? {
                val charArray = arg.toCharArray()
                if (charArray.size < 3 || charArray[0] != 'P' || charArray[1] != 'T') return null
                var tokenStart = 2
                var sec = 0.0
                for (i in 2 until charArray.size) {
                    if (charArray[i] == 'S' || charArray[i] == 'M' || charArray[i] == 'H' && tokenStart != i) {
                        val token = String(charArray, tokenStart, i - tokenStart)
                        val parseDouble = token.toDouble()
                        tokenStart = i + 1
                        when (charArray[i]) {
                            'S' -> sec += parseDouble
                            'M' -> sec += parseDouble * 60
                            'H' -> sec += parseDouble * 3600
                        }
                    }
                }
                return Time(sec)
            }
        }

    }

    class TypeHandler1 : TypeHandler {
        override fun parse(value: String, clz: Class<*>): Any? {
            return if (clz == Time::class.java) {
                Time.parseTime(value)!!
            } else null
        }

        override fun supports(clz: Class<*>): Boolean {
            return clz == Time::class.java
        }
    }

    class MPD {
        var mediaPresentationDuration: Time? = null
        var periods: MutableList<Period> = LinkedList()
        fun addPeriod(arg: Period) {
            periods.add(arg)
        }
    }

    class Period {
        var start: Time? = null
        var duration: Time? = null
        var adaptationSets: MutableList<AdaptationSet> = LinkedList()
        fun addAdaptationSet(arg: AdaptationSet) {
            adaptationSets.add(arg)
        }
    }

    class AdaptationSet {
        var segmentAlignment = false
        var maxWidth = 0
        var maxHeight = 0
        var maxFrameRate: Rational? = null
        var par: Rational? = null
        var segmentTemplate: SegmentTemplate? = null
        var representations: MutableList<Representation> = LinkedList()
        fun addRepresentation(arg: Representation) {
            representations.add(arg)
        }
    }

    class SegmentTemplate {
        var timescale = 0
        var duration = 0
        var media: String? = null
        var startNumber = 0
        var initialization: String? = null
    }

    class Representation {
        var id: String? = null
        var mimeType: String? = null
        var codecs: String? = null
        var width = 0
        var height = 0
        var frameRate: Rational? = null
        var sar: Rational? = null
        var startWithSAP = 0
        var bandwidth = 0
        var audioSamplingRate = 0
        var baseURL: String? = null
        var segmentTemplate: SegmentTemplate? = null
        var segmentBase: SegmentBase? = null
        var segmentList: SegmentList? = null
    }

    class SegmentBase {
        var indexRange // "0-834"
                : String? = null
    }

    class SegmentList {
        var duration = 0
        var initialization: Initialization? = null
        var segmentUrls: MutableList<SegmentURL> = LinkedList()
        fun addSegmentURL(arg: SegmentURL) {
            segmentUrls.add(arg)
        }
    }

    class Initialization {
        var sourceURL: String? = null
    }

    class SegmentURL {
        var media: String? = null
    }
}